File: 153.find-minimum-in-rotated-sorted-array.js
/*
 * @lc app=leetcode id=153 lang=javascript
 *
 * [153] Find Minimum in Rotated Sorted Array
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMin = function (nums) {
  let l = 0;
  let r = nums.length - 1;
  let res = nums[0];
  while (l <= r) {
    const m = parseInt(l + (r - l) / 2);
    res = Math.min(res, nums[l], nums[m]); // 这三个地方可能最小值
    if (nums[m] > nums[r]) {
      // right side is unsorted
      l = m + 1;
    } else if (nums[m] < nums[l]) {
      // left side is unsorted
      r = m - 1;
    } else {
      break;
    }
  }
  return res;
};
// @lc code=end

File: solutions/153.find-minimum-in-rotated-sorted-array.rb
# @param {Integer[]} nums
# @return {Integer}
def find_min(nums)
  i = 0
  j = nums.size - 1
  found = -1
  while i <= j #  实际上是不会跳出循环，当 left==right 时直接返回
    if nums[i] <= nums[j] # target 变种
      found = nums[i]
      break
    end

    m = i + (j - i) / 2
    if nums[i] <= nums[m]
      i = m + 1
    else
      j = m # 注意 right 更新时会被设为 mid 而不是 mid-1，因为 mid 无法被排除。
    end
  end
  found
end
